### PR TITLE
Pcre engine

### PR DESCRIPTION
**Pull request by connorchang**
This pull request adds the PCRE2 engine to the regexbench. This is the first step of PCRE engine implementation. There are two more, namely PCRE, PCRE2-JIT, coming after it gets merged. 

I have one issue needed to be fixed. When it gets compiled, the warning below showed up, and I am not sure hot to remove it.

```
/Users/connor_petabi/regexbench/src/PCRE2Engine.cpp:17:46: warning: disabled expansion of recursive macro [-Wdisabled-macro-expansion]
    auto re = pcre2_compile(reinterpret_cast<PCRE2_SPTR>(rule.getRegexp().data()),
                                             ^
/opt/pkg/include/pcre2.h:561:50: note: expanded from macro 'PCRE2_SPTR'
#define PCRE2_SPTR                  PCRE2_SUFFIX(PCRE2_SPTR)
```

### CODE REVIEW

File: src/PCRE2Engine.h [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+class PCRE2Engine : public Engine {
+public:
+  PCRE2Engine();
+  virtual ~PCRE2Engine();
+
+  virtual void compile(const std::vector<Rule> &);
+  virtual bool match(const char *, size_t);
+
+private:
+  std::vector<pcre2_code *>res;
```
Comment (msk) [inline src/PCRE2Engine.h]: Make this `vector<unique_ptr<pcre2_code, decltype(pcre2_code_free)>>` so that we don't have to call `pcre2_code_free()` explicitly in the destructor.

File: src/PCRE2Engine.cpp [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+bool PCRE2Engine::match(const char *data, size_t len) {
+  for (const auto &re : res) {
+    pcre_matching_data = pcre2_match_data_create_from_pattern(re, nullptr);
+    int rc = pcre2_match(re,
+                         reinterpret_cast<PCRE2_SPTR>(data),
+                         len, 0, PCRE2_NOTEMPTY_ATSTART |
+                         PCRE2_NOTEMPTY,
+                         pcre_matching_data, nullptr);
+    pcre2_match_data_free(pcre_matching_data);   /* Release memory used for the match */
+    if (rc < 0) {
```
Comment (msk) [inline src/PCRE2Engine.cpp]: Wouldn't the following `if` statement be easier to read?

``` c++
if (rc >= 0)
  return true;
```

File: src/PCRE2Engine.cpp [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+  if (rule.isSet(REMATCH_MOD_MULTILINE))
+    opt |= PCRE2_MULTILINE;
+  if (rule.isSet(REMATCH_MOD_DOTALL))
+    opt |= PCRE2_DOTALL;
+  return opt;
+}
+
+bool PCRE2Engine::match(const char *data, size_t len) {
+  for (const auto &re : res) {
+    pcre_matching_data = pcre2_match_data_create_from_pattern(re, nullptr);
```
Comment (msk) [inline src/PCRE2Engine.cpp]: We need to minimize the amount of work during matching, moving everything that can be done earlier to `::compile()`. As far as I understand, `pcre2_match()` never reads from its 6th argument, `match_data`, which means we can create the `pcre2_match_data` object for each pattern in `::compile()`.

File: src/regexbench.cpp [hunk 4]
```diff
@@ -12,12 +12,14 @@
 #include "regexbench.h"
 #include "REmatchEngine.h"
 #include "Rule.h"
+#include "PCRE2Engine.h"
```
Comment (msk) [inline src/regexbench.cpp]: Sort `#include`'s alphabetically.

File: src/regexbench.cpp [hunk 5]
```diff
@@ -12,12 +12,14 @@
...
 
 namespace po = boost::program_options;
 
 enum EngineType : uint64_t {
   ENGINE_HYPERSCAN,
-  ENGINE_REMATCH
+  ENGINE_REMATCH,
+  ENGINE_PCRE2
```
Comment (msk) [inline src/regexbench.cpp]: Sort these alphabetically unless there is a reason to maintain a certain order.

File: src/regexbench.cpp [hunk 6]
```diff
@@ -44,6 +46,9 @@ int main(int argc, const char *argv[]) {
     case ENGINE_REMATCH:
       engine = std::make_unique<regexbench::REmatchEngine>();
       break;
+    case ENGINE_PCRE2:
```
Comment (msk) [inline src/regexbench.cpp]: Sort `case`'s alphabetically unless there is more logical grouping of cases or the alphabetical ordering affects performance.

File: src/PCRE2Engine.cpp [hunk 7]
```diff
@@ -0,0 +1,46 @@
+#include <rematch/compile.h>
+#include "PCRE2Engine.h"
+
+using namespace regexbench;
+
+PCRE2Engine::PCRE2Engine() {}
+
+PCRE2Engine::~PCRE2Engine() {}
```
Comment (msk) [inline src/PCRE2Engine.cpp]: Use default constructor/destructor (`= default` in the class definition) rather than define empty ones.

File: CMakeLists.txt [hunk 8]
```diff
@@ -47,6 +47,10 @@ if(NOT CMAKE_CXX_FLAGS_RELEASE MATCHES "-O")
   set(CMAKE_CXX_FLAGS_RELEASE "${CMAKE_CXX_FLAGS_RELEASE} -O3")
 endif()
 
+if(APPLE)
+  set(CMAKE_PREFIX_PATH /opt/pkg)
```
Comment (msk) [inline CMakeLists.txt]: Why is this needed? Is your cmake in `/opt/pkg/bin`? If so, cmake should find files under `/opt/pkg`.

Comment (connorchang) [inline CMakeLists.txt]: I think it's because I installed cmake through homebrew, so it gives me some path issue. I just deleted it and now it works, I will fix it soon.
